### DISCUSSION TITLE
Vector Database: Embeddings Plugin

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#1189] [Ideas]
Now that ArcadeDB can handle vectors (look also at the recent support of vectors from Gremlin!), the big thing missing, IMHO, is a simple way to generate embeddings.

While this is usually not the database job, it would be SUPER handy having a plugin that with just a quick configuration, can generate the vectors for the embeddings automatically from the record content!

The best stack I found so far is:
- Deep Java Library (with native support for TensorFlow and PyTorch from Java)
- A model from Hugging Face already available in the "Zoo"

While installing Deep Java Library (DJL from now on) could be a breeze or a big pain, depending on the platform where you're running your server, it's still the best solution I tried from Java. Also, they have a nice integration with Hugging Face models where it's automatically downloaded and installed the first time.

Of course, the speed of generating the embeddings is variable and depends on your hw/sw configuration. In particular, if you have a CUDA GPU or not. 

We could provide an asynchronous update of the records as soon as the embeddings are generated. I still have to think to a proper way to do this, but this could be the default setting for most users.

The plugin should be in a different project under GitHub because very specific and with a lot of dependencies.

**Comment by hrstoyanov** [comment]
@lvca have you seen this one:

https://unum-cloud.github.io/usearch/


It has java bindings, quantization...

**Reply by lvca** [reply]
I wish I knew it 4 months ago! It would be interesting doing a benchmark with USearch and our HNSW index to see the actual difference.

**Comment by hrstoyanov** [comment]
@lvca these just keep coming ... this one might be the best one - no dependencies, quantization, and SIMD instructions/Panama API accelerator for Java 20+:

https://github.com/jbellis/jvector

**Reply by lvca** [reply]
Yep, we are already evaluating it. The code is well written, many test cases, it's already a good release! I'll find some time to do a POC with ArcadeDB on a separate branch and do some benchmarks.

**Reply by hrstoyanov** [reply]
Cool!...found some more info:
 https://foojay.io/today/jvector-1-0/
